### PR TITLE
Fixed WPA configuration and f2fs installation

### PR DESCRIPTION
**Pull request by nicholash1994**
(This is just a copy of the commit message in my fork of void-mklive)

The void-installer wifi configuration never worked for me. Whenever
I installed void, I always had to configure wpa_supplicant manually.
The reason is that the void-installer uses dhcpcd to hook into
wpa_supplicant, but the void-installer doesn't copy the wpa_supplicant hook
from /usr/share/dhcpcd/hooks into /usr/libexec/dhcpcd-hooks. After I modified
the installer to do this during wifi configuration, it worked! I also set it
up to copy the wpa_supplicant hook into /usr/libexec/dhcpcd-hooks in the
installation directory. So WI-FI should work out of the box now!

Another issue I had with the void installer was that I could never get f2fs
to work on void. This is because the kernel is stored in /boot, but the EFI
partition is mounted at /boot/efi. So if f2fs is used for the root filesystem,
then grub will return an unknown filesystem error since it doesn't support
f2fs.

To fix this, I modified the installer to allow the EFI system partition to
optionally be mounted at /boot instead of /boot/efi. After doing this, I
installed void on a flash drive with f2fs for the root partition and the
EFI system partition mounted at /boot and it worked fine! I also added the
-f flag to the mkfs.f2fs command in the create_filesystems() function. Before,
the installer wouldn't overwrite an existing f2fs partition, even if you
specified that you wanted a new filesystem to be created.

I also added support for nilfs2 in the installer. However, the installer
script doesn't add nilfs-utils into the live image by default, so you
have to specify it when running the installer script via the -p option.
Based on tests I ran, grub doesn't detect kernels on nilfs2 partitions,
even though the GRUB manual says that GRUB supports nilfs2.
So like f2fs, if you want to use nilfs2, the EFI system
partition has to be mounted at /boot.

Another issue I ran into was that the installer.sh script doesn't
automatically include dialog (the program the installer uses for the TUI)
in the live image. It has to be specified via the -p option.

However, I ran into one issue which I haven't fixed yet. When installing void
on a flash drive, everything went smoothly, but when I rebooted, my UEFI boot
manager failed to load the grub EFI binary. I had to reconfigure it in the
UEFI settings, after which it booted up fine. I ran into pretty much the same issue with
the official void installer ISO (the one on the website). So I don't think this bug was
caused by the changes I made to the installer.

### CODE REVIEW

**Comment by Vaelatern** [comment]
Should note that the dhcpcd hook has been depreciated. https://voidlinux.org/news/2019/01/dhcpcd-7.1.0.html

**Comment by nicholash1994** [comment]
I didn't know that. But when I had wpa_supplicant running as a separate service, it didn't work consistently. Sometimes when I booted up my computer, I would have to restart dhcpcd in order to get wifi to work. The dhcpcd hook is the only way I've been able to get wifi to work consistently

**Comment by nicholash1994** [comment]
I modified the installer to not use the deprecated wpa_supplicant hook. I tested it and everything worked perfectly.

File: installer.sh.in [hunk 1]
```diff
@@ -94,15 +94,15 @@ WIDGET_SIZE="10 70"
...
+        --backtitle "${BOLD}${WHITE}Void Linux installation -- https://www.voidlinux.org/ (0.22 616e813)${RESET}" \
         --cancel-label "Back" --aspect 20 "$@" 2>$ANSWER
     return $?
 }
 
 INFOBOX() {
     # Note: dialog --infobox and --keep-tite don't work together
     dialog --colors --no-shadow --no-mouse \
-        --backtitle "${BOLD}${WHITE}Void Linux installation -- https://www.voidlinux.org/ (@@MKLIVE_VERSION@@)${RESET}" \
+        --backtitle "${BOLD}${WHITE}Void Linux installation -- https://www.voidlinux.org/ (0.22 616e813)${RESET}" \
```
Comment (Vaelatern) [inline installer.sh.in]: Why is this hardcoded?

File: installer.sh.in [hunk 2]
```diff
@@ -493,10 +494,14 @@ at the first 2GB of the disk with the TOGGLE \`bios_grub' enabled.\n
 ${BOLD}NOTE: you don't need this on EFI systems.${RESET}\n\n
 For EFI systems GPT is mandatory and a FAT32 partition with at least\n
 100MB must be created with the TOGGLE \`boot', this will be used as\n
-EFI System Partition. This partition must have mountpoint as \`/boot/efi'.\n\n
+EFI System Partition. This partition must have mountpoint as /boot/efi\n
+or /boot.\n\n
```
Comment (Vaelatern) [inline installer.sh.in]: Wrong.

File: installer.sh.in [hunk 3]
```diff
@@ -493,10 +494,14 @@ at the first 2GB of the disk with the TOGGLE \`bios_grub' enabled.\n
...
 At least 1 partition is required for the rootfs (/).\n
 For swap, RAM*2 must be really enough. For / 600MB are required.\n\n
-${BOLD}WARNING: /usr is not supported as a separate partition.${RESET}\n
+${BOLD}WARNING: /usr is not supported as a separate partition.${RESET}\n\n
+${BOLD}WARNING:${RESET} If f2fs or nilfs2 is used as the root filesystem, then \n
+the boot partition MUST be mounted at /boot, and a filesystem must be used\n
+that GRUB supports, (i.e. btrfs, ext2/ext3/ext4, vfat, or xfs)${RESET}\n\n
```
Comment (Vaelatern) [inline installer.sh.in]: Why?

File: installer.sh.in [hunk 4]
```diff
@@ -837,7 +849,16 @@ configure_wifi() {
...
     rm -f ${_wpasupconf%.conf}-${dev}.conf
-    cp -f ${_wpasupconf} ${_wpasupconf%.conf}-${dev}.conf
+
+    # wpa_supplicant defaults
+    echo "ctrl_interface=/run/wpa_supplicant" >> ${_wpasupconf%.conf}-${dev}.conf
+    echo "ctrl_interface_group=wheel"         >> ${_wpasupconf%.conf}-${dev}.conf
+    echo "eapol_version=1"                    >> ${_wpasupconf%.conf}-${dev}.conf
+    echo "ap_scan=1"                          >> ${_wpasupconf%.conf}-${dev}.conf
+    echo "fast_reauth=1"                      >> ${_wpasupconf%.conf}-${dev}.conf
+    echo "update_config=1"                    >> ${_wpasupconf%.conf}-${dev}.conf
```
Comment (Vaelatern) [inline installer.sh.in]: Why hardcoded here?

File: installer.sh.in [hunk 5]
```diff
@@ -846,9 +867,32 @@ configure_wifi() {
         echo "  auth_alg=SHARED" >> ${_wpasupconf%.conf}-${dev}.conf
         echo "}" >> ${_wpasupconf%.conf}-${dev}.conf
     else
-        wpa_passphrase "$ssid" "$pass" >> ${_wpasupconf%.conf}-${dev}.conf
+        echo "network={" >> ${_wpasupconf%.conf}-${dev}.conf
+        echo "  ssid=\"$ssid\"" >> ${_wpasupconf%.conf}-${dev}.conf
+        echo "  scan_ssid=1" >> ${_wpasupconf%.conf}-${dev}.conf
+        echo "  key_mgmt=WPA-PSK" >> ${_wpasupconf%.conf}-${dev}.conf
+        echo "  psk=\"$pass\"" >> ${_wpasupconf%.conf}-${dev}.conf
+        echo "}" >> ${_wpasupconf%.conf}-${dev}.conf
```
Comment (Vaelatern) [inline installer.sh.in]: This.... this already is what is done. Why are you doing this?

File: installer.sh.in [hunk 6]
```diff
@@ -846,9 +867,32 @@ configure_wifi() {
...
+    sv check dhcpcd >$LOG 2>&1
+    if [ ! $? ]; then
+        DIALOG --msgbox "Error: dhcpcd couldn't be restarted!" ${MSGBOXSIZE}
+        return 1
+    else
+        INFOBOX "Setting up network..." 4 60
+    fi
+
+    # wait a couple seconds to setup wifi
+    sleep 5
```
Comment (Vaelatern) [inline installer.sh.in]: Racey and a bad idea.

File: installer.sh.in [hunk 7]
```diff
@@ -846,9 +867,32 @@ configure_wifi() {
...
+    ln -sf /etc/sv/dhcpcd /var/service
+    sv restart wpa_supplicant >$LOG 2>&1
+    sv restart dhcpcd >$LOG 2>&1
+    sv check dhcpcd >$LOG 2>&1
+    if [ ! $? ]; then
+        DIALOG --msgbox "Error: dhcpcd couldn't be restarted!" ${MSGBOXSIZE}
+        return 1
+    else
+        INFOBOX "Setting up network..." 4 60
+    fi
```
Comment (Vaelatern) [inline installer.sh.in]: This whole conditional is unreliable

File: installer.sh.in [hunk 8]
```diff
@@ -881,6 +925,10 @@ configure_net_dhcp() {
             DIALOG --msgbox "${BOLD}${RED}ERROR:${RESET} failed to run dhcpcd. See $LOG for details." ${MSGBOXSIZE}
             return 1
         fi
+
+        # wait a little bit for wifi to configure
+        sleep 10
+
```
Comment (Vaelatern) [inline installer.sh.in]: uhhhhh

File: installer.sh.in [hunk 9]
```diff
@@ -965,7 +1013,7 @@ validate_filesystems() {
             rootfound=1
         elif [ "$mntpt" = "/usr" ]; then
             usrfound=1
-        elif [ "$fstype" = "vfat" -a "$mntpt" = "/boot/efi" ]; then
+        elif [ "$fstype" = "vfat" -a \( "$mntpt" = "/boot/efi" -o "$mntpt" = "/boot" \) ]; then
```
Comment (Vaelatern) [inline installer.sh.in]: No. No no no no no.

File: installer.sh.in [hunk 10]
```diff
@@ -985,10 +1033,12 @@ the mount point for the root filesystem (/) has not yet been configured." ${MSGB
         DIALOG --msgbox "${BOLD}${RED}ERROR:${RESET} \
 /usr mount point has been configured but is not supported, please remove it to continue." ${MSGBOXSIZE}
         return 1
-    elif [ -n "$EFI_SYSTEM" -a -z "$efi_system_partition" ]; then
+    elif [ -n "$EFI_SYSTEM" -a \( -z "$efi_system_partition" \) ]; then
```
Comment (Vaelatern) [inline installer.sh.in]: Why?

File: installer.sh.in [hunk 11]
```diff
@@ -985,10 +1033,12 @@ the mount point for the root filesystem (/) has not yet been configured." ${MSGB
...
         DIALOG --msgbox "${BOLD}${RED}ERROR:${RESET} \
 The EFI System Partition has not yet been configured, please create it\n
-as FAT32, mountpoint /boot/efi and at least with 100MB of size." ${MSGBOXSIZE}
+as FAT32, mountpoint /boot/efi or /boot and at least with 100MB of size.\n
+If f2fs or nilfs2 is used on the root partition, then the boot partition MUST be\n
+mounted at /boot" ${MSGBOXSIZE}
```
Comment (Vaelatern) [inline installer.sh.in]: Why is this a requirement for these systems anyway?

File: installer.sh.in (hunk 2 above)
Comment (the-maldridge) [reply installer.sh.in]: The terminology is wrong here.  The one that bricks your system is efivars, this is just the EFI location for boot.

File: installer.sh.in [hunk 12]
```diff
@@ -1031,9 +1081,10 @@ failed to activate swap on $dev!\ncheck $LOG for errors." ${MSGBOXSIZE}
             ext2) MKFS="mke2fs -F"; modprobe ext2 >$LOG 2>&1;;
             ext3) MKFS="mke2fs -F -j"; modprobe ext3 >$LOG 2>&1;;
             ext4) MKFS="mke2fs -F -t ext4"; modprobe ext4 >$LOG 2>&1;;
-            f2fs) MKFS="mkfs.f2fs"; modprobe f2fs >$LOG 2>&1;;
+            f2fs) MKFS="mkfs.f2fs -f"; modprobe f2fs >$LOG 2>&1;;
```
Comment (Vaelatern) [inline installer.sh.in]: Why this change?

File: installer.sh.in [hunk 13]
```diff
@@ -1187,7 +1238,9 @@ install_packages() {
 }
 
 enable_dhcpd() {
-    ln -sf /etc/sv/dhcpcd $TARGETDIR/etc/runit/runsvdir/default/dhcpcd
+    ln -sf /etc/sv/dhcpcd $TARGETDIR/etc/runit/runsvdir/default
+    ln -sf /etc/sv/wpa_supplicant $TARGETDIR/etc/runit/runsvdir/default
+    cp -r /etc/wpa_supplicant/* $TARGETDIR/etc/wpa_supplicant
 }
```
Comment (Vaelatern) [inline installer.sh.in]: And if they don't want wifi?

File: installer.sh.in (hunk 5 above)
Comment (nicholash1994) [reply installer.sh.in]: The reason why I replaced "wpa_passphrase ...." by hardcoding it is because it's consistent with how WEP is configured.  I mean it's not a big deal, you can change that.

**Comment by nicholash1994** [comment]
Ok, so I will admit I'm not great at shell scripting, but I explained pretty much all the decisions I made in the initial commit message. I know it's long, but it answer most of the questions you asked. The reason why I want the option to mount the EFI partition at /boot, is because the kernel is stored in /boot. So if the EFI partition was mounted at /boot/efi, and the only other partition was the root partition, then the kernel will be stored on the root partition. Now this can be an issue because GRUB doesn't support f2fs, so if you want to use f2fs as your root partition, the installer will only work if you either mount the EFI partition at /boot, or make a /boot partition for the kernel in addition to a /boot/efi partition for GRUB. And making a separate /boot/efi and /boot  partition seems really unnecessary. It's a lot easier to just mount the EFI partition at /boot. I mean I tried it and it works fine, so I don't see what the big deal is.

The changes I made to the wifi configuration are arguably pretty bad, but it worked on my laptop at least. Before I made those changes, I could never get wifi to work with the installer. I don't care that much about the WIFI configuration because that's easy to do on the command-line. But I think it's really important to have the option to mount the EFI partition at /boot, because otherwise the GRUB installation fails if f2fs is used as the root partition (because GRUB doesn't support f2fs).

File: installer.sh.in (hunk 12 above)
Comment (nicholash1994) [reply installer.sh.in]: I also explained the reason for this in the initial commit message. The reason is that if you don't do that, if you ask the installer to create a new f2fs filesystem on an f2fs partition, then it won't work (because by default mkfs.f2fs won't overwrite an existing f2fs filesystem, like the mkfs for other filesystems). It's the same reason why the "-F" option is present for the other mkfs commands

File: installer.sh.in (hunk 10 above)
Comment (nicholash1994) [reply installer.sh.in]: Because before I made that change, that conditional would return TRUE, even if the $efi_system_partition variable was set

File: installer.sh.in (hunk 4 above)
Comment (nicholash1994) [reply installer.sh.in]: Because I want ${_wpasupconf} to be a symlink to ${_wpasupconf%.conf}-${dev}.conf. On the root filesystem, the default wpa_supplicant configuration is stored ${_wpasupconf}. When the installer runs, ${_wpasupconf} is copied to ${_wpasupconf%.conf}-${dev}.conf, and then options are appended to the latter file. However, if ${_wpasupconf} is used as symlink to the wpa_supplicant configuration file for the configured device, then the default wpa_supplicant options have to be stored somewhere else. I just hardcoded them in the installer, but one could easily just create /etc/wpa_supplicant/wpa_supplicant.defaults, and put the default options there.

File: installer.sh.in (hunk 13 above)
Comment (nicholash1994) [reply installer.sh.in]: True. wpa_supplicant should be enabled somewhere else in the installer.

File: installer.sh.in (hunk 9 above)
Comment (nicholash1994) [reply installer.sh.in]: Why?

**Comment by the-maldridge** [comment]
This PR is a trainwreck.  Please at minimum split this into two different PRs, one for your wpa_supplicant changes and one for the f2fs patchset.

**Comment by nicholash1994** [comment]
Okay, sorry.. I'll also create a separate PR for the f2fs patchset. But what about the option to mount the boot partition at /boot (instead of /boot/efi)? That's the only way I could get an f2fs installation to work.

**Comment by the-maldridge** [comment]
Put that in the f2fs PR and it can be discussed there.

File: installer.sh.in (hunk 5 above)
Comment (Vaelatern) [reply installer.sh.in]: I don't understand what you are saying.

File: installer.sh.in (hunk 4 above)
Comment (Vaelatern) [reply installer.sh.in]: I still don't understand why these options are hardcoded here.

File: installer.sh.in (hunk 5 above)
Comment (nicholash1994) [reply installer.sh.in]: Ok, so the way the WEP configuration is modified is like this

    echo "network={" >> ${_wpasupconf%.conf}-${dev}.conf
            echo "  ssid=\"$ssid\"" >> ${_wpasupconf%.conf}-${dev}.conf
            echo "  wep_key0=\"$pass\"" >> ${_wpasupconf%.conf}-${dev}.conf
            echo "  wep_tx_keyidx=0" >> ${_wpasupconf%.conf}-${dev}.conf
            echo "  auth_alg=SHARED" >> ${_wpasupconf%.conf}-${dev}.conf
    echo "}" >> ${_wpasupconf%.conf}-${dev}.conf

The configuration for WPA used to be this

    wpa_passphrase "$ssid" "$pass" >> ${_wpasupconf%.conf}-${dev}.conf

So I changed that to this

    echo "  scan_ssid=1" >> ${_wpasupconf%.conf}-${dev}.conf    
    echo "  key_mgmt=WPA-PSK" >> ${_wpasupconf%.conf}-${dev}.conf
    echo " psk=\"$pass\"" >> ${_wpasupconf%.conf}-${dev}.conf
    echo "}" >> ${_wpasupconf%.conf}-${dev}.conf

I mean this change is not a big deal, I don't care if it's merged or not. But the reason why I changed it is that it makes the WPA configuration consistent with the WEP configuration. Basically, it just looks nicer IMO. It doesn't do anything differently, I just think it makes the code somewhat more readable.

**Comment by the-maldridge** [comment]
This has been superseded by #46
